python-frontend/src/main/java/org/sonar/python/types/TypeShed.java

```diff
@@ -340,4 +354,77 @@ private static void setParameterTypes(Symbol symbol, FunctionDef functionDef) {
+  }
+
+  @CheckForNull
+  static ModuleSymbol deserializedModule(String moduleName, InputStream resource) {
+    try {
+      return ModuleSymbol.parseFrom(resource);
+    } catch (IOException e) {
+      LOG.debug("Error while deserializing protobuf for module " + moduleName, e);
+      return null;
+    }
+  }
+
+  static Map<String, Symbol> getSymbolsFromProtobufModule(@Nullable ModuleSymbol moduleSymbol) {
+    if (moduleSymbol == null) {
+      return Collections.emptyMap();
+    }
+    Map<String, Symbol> deserializedSymbols = new HashMap<>();
+    for (SymbolsProtos.ClassSymbol classSymbolProto : moduleSymbol.getClassesList()) {
+      ClassSymbolImpl classSymbol = new ClassSymbolImpl(classSymbolProto);
+      addSuperClasses(deserializedSymbols, classSymbolProto, classSymbol);
+      deserializedSymbols.put(classSymbolProto.getFullyQualifiedName(), classSymbol);
+    }
+    for (SymbolsProtos.FunctionSymbol functionSymbolProto : moduleSymbol.getFunctionsList()) {
+      deserializedSymbols.put(functionSymbolProto.getFullyQualifiedName(), new FunctionSymbolImpl(functionSymbolProto));
+    }
+    for (SymbolsProtos.OverloadedFunctionSymbol overloadedFunctionSymbol : moduleSymbol.getOverloadedFunctionsList()) {
+      deserializedSymbols.put(overloadedFunctionSymbol.getFullname(),
+        AmbiguousSymbolImpl.create(overloadedFunctionSymbol.getDefinitionsList().stream().map(FunctionSymbolImpl::new).collect(Collectors.toSet())));
+    }
+    return deserializedSymbols;
+  }
+
+
+  private static void addSuperClasses(Map<String, Symbol> deserializedSymbols, SymbolsProtos.ClassSymbol classSymbolProto, ClassSymbolImpl classSymbol) {
+    classSymbolProto.getSuperClassesList().forEach(superClassFqn -> {
+      if (deserializedSymbols.containsKey(superClassFqn)) {
+        classSymbol.addSuperClass(deserializedSymbols.get(superClassFqn));
+        return;
+      }
+      String[] fqnSplittedByDot = superClassFqn.split("\\.");
+      String localName = fqnSplittedByDot[fqnSplittedByDot.length - 1];
+      Symbol symbol = symbolWithFQN(superClassFqn);
+      if (symbol != null) {
+        classSymbol.addSuperClass(symbol);
+        return;
+      }
+      classSymbol.addSuperClass(new SymbolImpl(localName, superClassFqn));
+    });
+  }
+
+  @CheckForNull
+  public static Symbol symbolWithFQN(String fullyQualifiedName) {
+    String[] fqnSplittedByDot = fullyQualifiedName.split("\\.");
+    String localName = fqnSplittedByDot[fqnSplittedByDot.length - 1];
+    if (fqnSplittedByDot.length == 2 && fqnSplittedByDot[0].equals("builtins") && builtins.containsKey(localName)) {
+      return builtins.get(localName);
+    }
+    for (Map<String, Symbol> symbolsByFqn : typeShedSymbols.values()) {
+      Symbol symbol = symbolsByFqn.get(fullyQualifiedName);
+      if (symbol != null) {
+        return symbol;
+      }
+    }
+    return builtins.getOrDefault(fullyQualifiedName, null);
```

Comment:
Can it happen that we actually retrieve a builtin here if the name didn't start with `builtins`? (for serialized Typeshed at least)

Reply:
I think you're right, it should never happen for serialized Typeshed